README.md
TracMovieMacro
==============

This repository is a clone of the Trac-Hacks SVN repository.

http://trac-hacks.org/svn/moviemacro

see also: http://trac-hacks.org/wiki/MovieMacro


### Tips

To fetch/rebase from original svn repository:

    git config --add svn-remote.svn.url http://trac-hacks.org/svn/moviemacro
    git config --add svn-remote.svn.fetch :refs/remotes/git-svn
    git checkout -b mylocalbranch
    git svn fetch svn
    git svn rebase svn

To get original source code after fetch remote svn repository:

    git checkout -b origlocalbranch git-svn


trunk/CHANGELOG

ChangeLog
=========

0.3 (2014-11-05)
------------------

* Change license from BSD to Apache License 2.0
* Update Flowplayer 5.5.0
  - Support to play mp4/webm format
* Refactoring to be able to maintain

0.2 (2014-10-30)
------------------

* Initial PyPI package as a checkpoint


trunk/tests/conftest.py
# -*- coding: utf-8 -*-
from trac.attachment import AttachmentModule
from trac.resource import ResourceSystem
from trac.test import EnvironmentStub

from movie.macro import MovieMacro
from movie.web_ui import MoviePreviewRenderer


def pytest_addoption(parser):
    group = parser.getgroup("general")
    group.addoption('--envscope',
                    action="store", dest="envscope", default="module",
                    type="choice", choices=["module", "function"],
                    help=("set environment scope, default: module."))


def make_trac_environment_with_plugin():
    env = EnvironmentStub(enable=["movie.*", MovieMacro, MoviePreviewRenderer])
    resource_system = ResourceSystem(env)
    resource_system._resource_managers_map = {
        'attachment': AttachmentModule(env),
    }
    movie_macro = MovieMacro(env)
    preview_renderer = MoviePreviewRenderer(env)
    return env, movie_macro, preview_renderer


def pytest_funcarg__env(request):
    setup = make_trac_environment_with_plugin
    scope = request.config.option.envscope
    env, _, _ = request.cached_setup(setup=setup, scope=scope)
    return env


def pytest_funcarg__movie_macro(request):
    setup = make_trac_environment_with_plugin
    scope = request.config.option.envscope
    _, movie_macro, _ = request.cached_setup(setup=setup, scope=scope)
    return movie_macro


def pytest_funcarg__preview_rendere(request):
    setup = make_trac_environment_with_plugin
    scope = request.config.option.envscope
    _, _, preview_renderer = request.cached_setup(setup=setup, scope=scope)
    return preview_renderer


trunk/tests/test_macro.py
# -*- coding: utf-8 -*-
import pytest

from trac.web.href import Href


BASE_HREF = 'http://example.com/mysite'


class Request(object):
    locale = None
    args = {}

    def __init__(self, **kwargs):
        self.chrome = {}
        for k, v in kwargs.items():
            setattr(self, k, v)


@pytest.mark.parametrize(('url', 'expected'), [
    ('http://example.com/file.ext', 'http://example.com/file.ext'),
    ('htdocs://site/test.flv', '%s/chrome/site/test.flv' % BASE_HREF),
    ('chrome://site/test.flv', '%s/chrome/site/test.flv' % BASE_HREF),
    ('ticket://123/sample.webm',
     '%s/raw-attachment/ticket/123/sample.webm' % BASE_HREF),
    ('wiki://page/sample.mp4',
     '%s/raw-attachment/wiki/page/sample.mp4' % BASE_HREF),
    ('source://repo/movie.ogv', '%s/export/repo/movie.ogv' % BASE_HREF),
])
def test_get_absolute_url(movie_macro, url, expected):
    req = Request(abs_href=Href(BASE_HREF))
    assert expected == movie_macro._get_absolute_url(req, url)


@pytest.mark.parametrize(('url', 'path_info', 'expected'), [
    ('sample.webm', u'/ticket/123',
     '%s/raw-attachment/ticket/123/sample.webm' % BASE_HREF),
    ('ticket:123:sample.webm', u'/ticket/123',
     '%s/raw-attachment/ticket/123/sample.webm' % BASE_HREF),
    ('ticket:456:sample.webm', u'/ticket/123',
     '%s/raw-attachment/ticket/456/sample.webm' % BASE_HREF),
    ('wiki:sample.mp4', u'/wiki/page',
     '%s/raw-attachment/wiki/page/sample.mp4' % BASE_HREF),
    ('wiki:page/sample.mp4', u'/wiki/page',
     '%s/raw-attachment/wiki/page/sample.mp4' % BASE_HREF),
    ('wiki:test/movie/sub/sample.mp4', u'/wiki/test/movie/sub',
     '%s/raw-attachment/wiki/test/movie/sub/sample.mp4' % BASE_HREF),
])
def test_get_absolute_url_simple(movie_macro, url, path_info, expected):
    req = Request(abs_href=Href(BASE_HREF), path_info=path_info)
    assert expected == movie_macro._get_absolute_url(req, url)


trunk/tests/test_utils.py
# -*- coding: utf-8 -*-
import pytest

from movie.utils import _xform_to_dict
from movie.utils import parse_imagemacro_style
from movie.utils import set_default_parameters
from movie.utils import string_keys
from movie.utils import xform_query
from movie.utils import xform_style


@pytest.mark.parametrize(('d', 'expected'), [
    ({}, {}),
    ({u'x': 1, u'y': 2}, {'x': 1, 'y': 2}),
])
def test_string_keys(d, expected):
    assert expected == string_keys(d)


@pytest.mark.parametrize(('style', 'first_sep', 'second_sep', 'expected'), [
    ('display: none; clear: both;', ';', ':',
     {'display': 'none', 'clear': 'both'}),
])
def test_xform_to_dict(style, first_sep, second_sep, expected):
    assert expected == _xform_to_dict(style, first_sep, second_sep)


@pytest.mark.parametrize(('style', 'expected'), [
    ({}, ''),
    ('', {}),
    ({'display': 'none'}, 'display: none;'),
    ('clear: both;', {'clear': 'both'}),
])
def test_xform_style_simple(style, expected):
    assert expected == xform_style(style)


@pytest.mark.parametrize('style', [
    {'width': u'320px', 'height': u'240px', 'display': 'none'},
])
def test_xform_style_multi(style):
    assert style == xform_style(xform_style(style))


@pytest.mark.parametrize(('query', 'expected'), [
    ({}, ''),
    ('', {}),
    ({u'v': u'9dfWzp7rYR4'}, u'v=9dfWzp7rYR4'),
    ('v=9dfWzp7rYR4', {'v': '9dfWzp7rYR4'}),
])
def test_xform_query_simple(query, expected):
    assert expected == xform_query(query)


@pytest.mark.parametrize('query', [
    {'t': '1414735598', 'key': 'bW92aWVtYWNybw=='}
])
def test_xform_query_multi(query):
    assert query == xform_query(xform_query(query))


@pytest.mark.parametrize(('param', 'default', 'kwargs', 'expected'), [
    (
        {'width': '320', 'test': '1'},
        {'width': '640', 'height': '360', 'test': '0', 'p': None},
        {'test': '2'},
        {'width': '320', 'height': '360', 'test': '2'}
    ),
])
def test_set_default_parameters(param, default, kwargs, expected):
    set_default_parameters(param, default, **kwargs)
    assert expected == param


@pytest.mark.parametrize(('url', 'path_info', 'expected'), [
    ('file.ext', '/ticket/1', ('ticket', '1', 'file.ext')),
    ('ticket:1:file.ext', '/ticket/1', ('ticket', '1', 'file.ext')),
    ('ticket:2:file.ext', '/ticket/1', ('ticket', '2', 'file.ext')),
    ('file.ext', '/wiki/start', ('wiki', 'start', 'file.ext')),
    ('file.ext', '/wiki/start/sub/deep',
        ('wiki', 'start', 'sub/deep/file.ext')),
    ('wiki:start/sub/deep/file.ext', '/wiki/start/sub/deep',
        ('wiki', 'start', 'sub/deep/file.ext')),
    ('wiki:start/file.ext', '/wiki/start', ('wiki', 'start', 'file.ext')),
])
def test_parse_imagemacro_style(url, path_info, expected):
    assert expected == parse_imagemacro_style(url, path_info)


trunk/movie/macro.py
# -*- coding: utf-8 -*-
"""
    Movie plugin for Trac.

    Embeds various online movies.
"""
import mimetypes
from datetime import datetime
from os.path import join as pathjoin
from urlparse import urlparse

from genshi.builder import tag
from pkg_resources import resource_filename
from trac.core import TracError
from trac.core import implements
from trac.resource import Resource, get_resource_url
from trac.util.presentation import to_json
from trac.web.api import IRequestFilter
from trac.web.chrome import ITemplateProvider, add_script, add_stylesheet
from trac.wiki.api import parse_args
from trac.wiki.macros import WikiMacroBase

from model import MovieMacroConfig
from utils import parse_imagemacro_style
from utils import set_default_parameters
from utils import string_keys
from utils import xform_query
from utils import xform_style
from video_sites import get_embed_video_site_player


EMBED_PATH_FLOWPLAYER = {
    'js': 'movie/js/flowplayer.min.js',
    'css': 'movie/js/skin/minimalist.css',
    'swf': 'movie/swf/flowplayer.swf',
}

_EMBED_FLOWPLAYER_DEFAULT_PARAMETERS = {
    'adaptiveRatio': True,
    'bufferTime': '0.1',
    'debug': False,
    'disabled': False,
    'engine': 'html5',
    'flashfit': False,
    'fullscreen': True,
    'errors': 'array',
    'keyboard': True,
    'live': False,
    'muted': False,
    'native_fullscreen': False,
    'preload': None,
    'ratio': None,  # use adaptiveRatio by default
    'rtmp': None,
    'speeds': [0.25, 0.5, 1, 1.5, 2],
    'swf': None,  # set swf path on server-side
    'splash': False,
    'subscribe': False,
    'tooltip': True,
    'volume': '1',
}


class MovieMacro(WikiMacroBase):
    """ Embed online movies from YouTube, GoogleVideo and MetaCafe, and local
        movies via FlowPlayer.
    """
    implements(IRequestFilter, ITemplateProvider)

    # IWikiMacroProvider methods
    def expand_macro(self, formatter, name, content):
        args, kwargs = parse_args(content, strict=True)
        if len(args) == 0:
            raise TracError('URL to a movie at least required.')

        url = self._get_absolute_url(formatter.req, args[0])
        scheme, netloc, path, params, query, fragment = urlparse(url)

        try:
            style_dict = xform_style(string_keys(kwargs).get('style', ''))
        except Exception as e:
            raise TracError('Double check the `style` argument: ' + str(e))

        self.log.debug('moviemacro style_dict: ' + repr(style_dict))
        config = MovieMacroConfig(self.env, self.config)
        style = {
            'width': style_dict.get('width', config.width),
            'height': style_dict.get('height', config.height),
            'border': style_dict.get('border', 'none'),
            'margin': style_dict.get('margin', '0 auto'),
            'display': 'block',
            'clear': 'both'
        }
        self.log.debug('moviemacro style: ' + repr(style))

        site_player = get_embed_video_site_player(netloc)
        if site_player is not None:
            return site_player(scheme, netloc, path, query, style)

        parse_result = urlparse(args[0])
        query = parse_result.query
        if config.splash:
            splash_url = pathjoin(formatter.href.chrome(), config.splash)
            splash_style = 'background-color:#777; '\
                           'background-image:url(%s);' % splash_url
            style['style'] = splash_style
            query += '&splash=true'
        return self.embed_player(formatter, url, query, style)

    def embed_player(self, formatter, url, query, style):
        query_dict = xform_query(query)
        set_default_parameters(
            query_dict,
            _EMBED_FLOWPLAYER_DEFAULT_PARAMETERS
        )

        player_id = self._generate_player_id()
        swf = pathjoin(formatter.href.chrome(), EMBED_PATH_FLOWPLAYER['swf'])
        style.pop('width')  # use adaptiveRatio for player-size
        style.pop('height')  # use adaptiveRatio for player-size
        attrs = {
            'id': player_id,
            'data-swf': swf,
            'style': xform_style(style),
        }
        return tag.div(
            tag.video([
                tag.source(type=mimetypes.guess_type(url)[0], src=url),
                tag.script("""
                    $(function() {
                        $('#%s').flowplayer(%s);
                    });
                """ % (player_id, to_json(query_dict))
                ),
            ]),
            **attrs
        )

    # IRequestFilter methods
    def pre_process_request(self, req, handler):
        return handler

    def post_process_request(self, req, template, data, content_type):
        path = req.path_info
        if path.startswith('/ticket/') or path.startswith('/wiki') \
           or path.startswith('/attachment/'):
            add_script(req, EMBED_PATH_FLOWPLAYER['js'])
            add_stylesheet(req, EMBED_PATH_FLOWPLAYER['css'])
        return template, data, content_type

    # ITemplateProvider methods
    def get_htdocs_dirs(self):
        yield ('movie', resource_filename('movie', 'htdocs'))

    def get_templates_dirs(self):
        return []

    # Private methods

    def _generate_player_id(self):
        now = datetime.now()
        return 'player-%s-%s' % (now.strftime('%s'), now.microsecond)

    def _get_absolute_url(self, req, url):
        """ Generate an absolute url from the url with the special schemes
        {htdocs,chrome,ticket,wiki,source} simply return the url if given with
        {http,https,ftp} schemes.

        Also support restricted ImageMacro format style.
        It can be specified only filename on ticket or wiki page.

        Examples:
            http://example.com/filename.ext
                ie. http://www.google.com/logo.jpg

            chrome://site/filename.ext
            htdocs://img/filename.ext
            htdocs:/img/filename.ext
                note: `chrome` is an alias for `htdocs`

            ticket://123/specification.pdf
            ticket:123:specification.pdf

            wiki://WikiWord/attachment.jpg
            wiki:WikiWord/attachment.jpg

            source://1024/path/filename.ext
        """
        if '://' in url:
            scheme, netloc, path, query, params, fragment = urlparse(url)
        else:  # suppose ImageMacro style
            scheme, netloc, path = parse_imagemacro_style(url, req.path_info)

        if scheme in ('htdocs', 'chrome'):
            return req.abs_href.chrome(netloc + path)

        if scheme in ('source',):
            return req.abs_href.export(netloc + path)

        if scheme in ('wiki', 'ticket'):
            resource = Resource(scheme, netloc).child('attachment', path)
            kwargs = {'format': 'raw'}
            return get_resource_url(self.env, resource, req.abs_href, **kwargs)

        return url


trunk/movie/web_ui.py
# -*- coding: utf-8 -*-
import re
import urllib

from trac.core import Component, implements
from trac.mimeview.api import IHTMLPreviewRenderer
from trac.wiki.formatter import format_to_html


PREVIEW_PATH = r'/.*?/raw-attachment/(?P<realm>.*?)/(?P<path>.*)$'


class MoviePreviewRenderer(Component):

    implements(IHTMLPreviewRenderer)

    ENCODING = 'utf-8'
    MACRO_FORMAT = '[[Movie(%(realm)s://%(path)s)]]'
    MIMETYPE_PREFIX = 'video/'
    PREVIEW_PATTERN = re.compile(PREVIEW_PATH)

    # IHTMLPreviewRenderer methods

    def get_quality_ratio(self, mimetype):
        if mimetype.startswith(self.MIMETYPE_PREFIX):
            return 9
        return 0

    def render(self, context, mimetype, content, filename=None, url=None):
        if url is not None:
            match = re.match(self.PREVIEW_PATTERN, urllib.unquote(str(url)))
            if match:
                path_info = match.groupdict()
                macro = unicode(self.MACRO_FORMAT % path_info, self.ENCODING)
                return format_to_html(self.env, context, macro)
        return None


trunk/movie/video_sites.py
# -*- coding: utf-8 -*-
"""
    Support Video Sharing Sites

    Define embed player tag and set VIDEO_SHARING_SITES
    to resolve site name and player function
"""
import mimetypes
from urllib import urlencode
from urlparse import urlunparse

from genshi.builder import tag
from trac.core import TracError

from utils import set_default_parameters, xform_style, xform_query

SWF_MIME_TYPE = mimetypes.types_map.get('.swf')


def get_embed_video_site_player(netloc):
    for sites, func in VIDEO_SHARING_SITES.items():
        if netloc in sites:
            return func
    return None


_EMBED_YOUTUBE_DEFAULT_PARAMETERS = {
    'autohide': '2',
    'autoplay': '0',
    'cc_load_policy': None,
    'color': 'red',
    'controls': '1',
    'disablekb': '0',
    'enablejsapi': '0',
    'end': None,
    'fs': '1',
    'hl': None,
    'iv_load_policy': '1',
    'list': None,
    'listType': None,
    'loop': '0',
    'modestbranding': '1',
    'origin': None,
    'playerapiid': None,
    'playlist': None,
    'playsinline': None,
    'rel': '0',
    'showinfo': '0',
    'start': None,
    'theme': 'dark',
}


def embed_youtube(scheme, netloc, path, query, style):
    """
    Embed youtube player using iframe.
    _EMBED_YOUTUBE_DEFAULT_PARAMETERS is set to the embed player.

    see also: https://developers.google.com/youtube/player_parameters
    """
    query_dict = xform_query(query)
    set_default_parameters(query_dict, _EMBED_YOUTUBE_DEFAULT_PARAMETERS)
    video_id = query_dict.pop('v')
    query = urlencode(query_dict.items())
    url = urlunparse((scheme, netloc, '/embed/%s' % video_id, '', query, ''))
    return tag.iframe(src=url, style=xform_style(style))


def embed_metacafe(scheme, netloc, path, query, style):
    parts = filter(None, path.split('/'))
    try:
        path = '/embed/%s/' % parts[1]
    except:
        msg = "Non-standard URL, "\
              "don't know how to process it, file a ticket please."
        raise TracError(msg)
    return tag.iframe(
        src=urlunparse((scheme, netloc, path, '', '', '')),
        allowFullScreen='true',
        frameborder=0,
        width=style['width'],
        height=style['height'],
        style=xform_style(style)
    )


def embed_vimeo(scheme, netloc, path, query, style):
    parts = filter(None, path.split('/'))
    path = '/moogaloop.swf?clip_id=%s&amp;server=vimeo.com&amp;'\
           'show_title=1&amp;show_byline=1&amp;'\
           'show_portrait=0&amp;color=&amp;fullscreen=1' % parts[0]
    url = urlunparse((scheme, netloc, path, '', '', ''))
    return tag.object(
        tag.param(name='movie', value=url),
        tag.param(name='allowfullscreen', value='true'),
        tag.param(name='allowscriptaccess', value='always'),
        tag.embed(
            src=url,
            type=SWF_MIME_TYPE,
            allowfullscreen='true',
            allowscriptaccess='always',
            width=style['width'],
            height=style['height']
        ),
        style=xform_style(style)
    )


VIDEO_SHARING_SITES = {
    ('www.youtube.com', 'www.youtube-nocookie.com'): embed_youtube,
    ('www.metacafe.com',): embed_metacafe,
    ('vimeo.com', 'www.vimeo.com'): embed_vimeo,
}


trunk/movie/utils.py
# -*- coding: utf-8 -*-
from os.path import join as pathjoin


def string_keys(d):
    """ Convert unicode keys into string keys, suiable for func(**d) use.
    >>> {} == string_keys({})
    True
    >>> {'x': 1, 'y': 2} == string_keys({u'x': 1, u'y': 2})
    True
    """
    return dict((str(key), value) for key, value in d.items())


def _xform_to_dict(style, first_sep, second_sep):
    """
    >>> d = {'display': 'none', 'clear': 'both'}
    >>> d == _xform_to_dict('display: none; clear: both;', ';', ':')
    True
    """
    return dict((s.strip() for s in i.split(second_sep, 1))
                for i in filter(None, style.split(first_sep)))


def xform_style(style):
    """ Convert between a style-string and a style-dictionary.
    >>> xform_style({})
    ''
    >>> xform_style('')
    {}
    >>> xform_style({'display': 'none'})
    'display: none;'
    >>> xform_style('clear: both;')
    {'clear': 'both'}
    >>> style = {'width': u'320px', 'height': u'240px', 'display': 'none'}
    >>> style == xform_style(xform_style(style))
    True
    """
    if isinstance(style, dict):
        result = '; '.join(['%s: %s' % (k, v) for k, v in style.items()])
        if result:
            result += ';'
        return result
    else:
        return _xform_to_dict(style, ';', ':')


def xform_query(query):
    """ Convert between a query-string and a query-dictionary.
    >>> xform_query({})
    ''
    >>> xform_query('')
    {}
    >>> xform_query({u'v': u'9dfWzp7rYR4'})
    u'v=9dfWzp7rYR4'
    >>> xform_query('v=9dfWzp7rYR4')
    {'v': '9dfWzp7rYR4'}
    >>> query = {'t': '1414735598', 'key': 'bW92aWVtYWNybw=='}
    >>> query == xform_query(xform_query(query))
    True
    """
    if isinstance(query, dict):
        return '&'.join(['%s=%s' % (k, v) for k, v in query.items()])
    else:
        return _xform_to_dict(query, '&', '=')


def set_default_parameters(param, default, **kwargs):
    """
    >>> param = {'width': '320', 'test': '1'}
    >>> default = {'width': '640', 'height': '360', 'test': '0', 'p': None}
    >>> set_default_parameters(param, default, test='2')
    >>> sorted(param.items())
    [('height', '360'), ('test', '2'), ('width', '320')]
    """
    for key, value in default.items():
        if param.get(key) is None:
            if value is not None:
                param[key] = value
    if kwargs:
        for key, value in kwargs.items():
            param[key] = value


def parse_imagemacro_style(url, path_info):
    """
    >>> parse_imagemacro_style(u'file.ext', u'/ticket/1')
    (u'ticket', u'1', u'file.ext')
    >>> parse_imagemacro_style(u'ticket:1:file.ext', u'/ticket/1')
    (u'ticket', u'1', u'file.ext')
    >>> parse_imagemacro_style(u'ticket:2:file.ext', u'/ticket/1')
    (u'ticket', u'2', u'file.ext')
    >>> parse_imagemacro_style(u'file.ext', u'/wiki/start')
    (u'wiki', u'start', u'file.ext')
    >>> parse_imagemacro_style(u'file.ext', u'/wiki/start/sub/deep')
    (u'wiki', u'start', u'sub/deep/file.ext')
    >>> url = u'wiki:start/sub/deep/file.ext'
    >>> parse_imagemacro_style(url, u'/wiki/start/sub/deep')
    (u'wiki', u'start', u'sub/deep/file.ext')
    >>> parse_imagemacro_style(u'wiki:start/file.ext', u'/wiki/start')
    (u'wiki', u'start', u'file.ext')
    """
    _path, netloc = '', ''
    _path_info = filter(None, path_info.split('/'))
    scheme = _path_info[0]
    if len(_path_info) > 1:
        netloc = _path_info[1]
        _path = '/'.join(_path_info[2:])

    _url = url.split(':')
    if scheme == 'ticket' and len(_url) == 3:
        netloc = _url[1]

    path = _url[-1]
    if scheme != 'ticket':
        if len(_url) == 1:
            path = pathjoin(_path, path)
        else:
            path = path.split('/', 1)[-1]

    return scheme, netloc, path
